Request caching when offline

gouravkmar opened the pull request:
Issue Link :link:
No specific issue was raised for this feature; this PR introduces a new functionality for caching failed requests and retrying them when the network becomes available.

Goals :soccer:
Automatically cache failed network requests and retry them once the network is reachable.
Improve handling of network connectivity issues, reducing the need for manual retries.

Implementation Details :construction:
A new CachedRequestManager class is introduced to cache failed requests and retry them when the network is reachable.
An extension on DataRequest adds a cacheAndRetry() method to handle caching and automatic retries of failed requests.

Usage :rocket:
Simply use the cacheAndRetry() method when making requests with Alamofire:

AF.request("https://example.com/api").cacheAndRetry()

Testing Details :mag:
Unit tests ensure failed requests are cached, retried, and cleared after successful retries.
Manual testing verifies that the retry mechanism works when toggling network availability.

jshier commented:
Thanks for the PR, we've wanted something like this! However, there's no need for custom caching here, Alamofire's retry feature already handles that. All you need to do here is create a custom `RequestRetrier`-conforming type and handle the retry logic in there, Alamofire can asynchronously wait to know whether to retry. 

Additionally, given the [imminent deprecation of `NetworkReachabilityMonitor`](https://github.com/Alamofire/Alamofire/pull/3947), you should rebuild on top of `NWPathMonitor`. That will ensure it works well into the future.

With those changes I can do a more detailed review and we can get this merged.

gouravkmar commented:
> Thanks for the PR, we've wanted something like this! However, there's no need for custom caching here, Alamofire's retry feature already handles that. All you need to do here is create a custom `RequestRetrier`-conforming type and handle the retry logic in there, Alamofire can asynchronously wait to know whether to retry.
> 
> Additionally, given the [imminent deprecation of `NetworkReachabilityManager`](https://github.com/Alamofire/Alamofire/pull/3947), you should rebuild on top of `NWPathMonitor`. That will ensure it works well into the future.
> 
> With those changes I can do a more detailed review and we can get this merged.

I've added the support of multiple failed requests in case of session wide retrier and removed the custom caching flow, please do check

gouravkmar commented:
@marcprux can you please help merge this in case this is fine?
